File: python/anyfft_array.py
# -*- coding: utf-8 -*-
####################################################################################################
# Fast Fourier Transform -- Python 3 Version
# This version implements Cooley-Tukey algorithm for composite numbers (not powers of 2 only).
#
# José Alexandre Nalon
####################################################################################################
# Since Python is an interpreted language, all you have to do is to invoque the interpreter to run
# this program:
#
# $ python3 anyfft_array.py


####################################################################################################
# Import needed modules:
from numpy import *                            # Deal with arrays;
from time import perf_counter                  # Time events;


####################################################################################################
# Definitions:
REPEAT = 50                                    # Number of executions to compute average time;


####################################################################################################
# Auxiliary Function:
def time_it(f, size, repeat=REPEAT):
    """
    Measure execution time through repeated calls to a (Fast) Fourier Transform function.

    :Parameters:
      f
        Function to be called;
      size
        Size of the vector on which the transform will be applied;
      repeat
        Number of times the function will be called. Defaults to REPEAT.

    :Returns:
      The average execution time for that function with a vector of the given size.
    """
    x = arange(size, dtype=complex)            # Generate a vector;
    t0 = perf_counter()                        # Start a timer;
    for j in range(0, repeat):                 # Repeated calls;
        f(x)
    return (perf_counter() - t0) / repeat      # Compute average;


####################################################################################################
# Direct FT:
def direct_ft(x):
    """
    Discrete Fourier Transform directly from the definition, an algorithm that has O(N^2)
    complexity. This implementation uses NumPy arrays for conciseness.

    :Parameters:
      x
        The vector of which the DFT will be computed. Given the nature of the implementation, there
        is no restriction on the size of the vector.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of the vector;
    X = zeros(x.shape, dtype=complex)          # Accumulate the results;
    W = exp(-2j*pi/N)                          # Twiddle factors;
    Wk = 1.
    for k in range(0, N):                      # Compute the kth coefficient;
        Wkn = 1.
        for n in range(0, N):                  #   Operate the summation;
            X[k] = X[k] + x[n]*Wkn             #     Compute every term;
            Wkn = Wkn * Wk                     # Update twiddle factors;
        Wk = Wk * W
    return X


####################################################################################################
# Auxiliary function:
def __factor(n):
    """
    Smallest prime factor of a given number. If the argument is prime itself, then it is the return
    value.

    :Parameters:
      n
        Number to be inspected.

    :Returns:
      The smallest prime factor, or the number itself if it is already a prime.
    """
    rn = int(ceil(sqrt(n)))                    # Search up to the square root of the number;
    for i in range(2, rn+1):
        if n%i == 0:                           # When remainder is zero, factor is found;
            return i
    return n


####################################################################################################
# Recursive FFT:
def recursive_fft(x):
    """
    Fast Fourier Transform using a recursive decimation in time algorithm. This has smaller
    complexity than the direct FT, though the exact value is difficult to compute. This
    implementation uses NumPy arrays for conciseness.

    :Parameters:
      x
        The vector of which the FFT will be computed. It must be a composite number, or else the
        computation will be defered to the direct FT, and there will be no efficiency gain.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of the vector;
    N1 = __factor(N)                           # Find the smallest factor of the vector length;
    if N1 == N:                                # If the length is prime itself,
        return direct_ft(x)                    #    the transform is given by the direct form;
    else:
        N2 = N // N1                           # Decompose in two factors, N1 being prime;
        X = zeros((N, ), dtype=complex)        # Accumulate the results;
        W = exp(-2j*pi/N)                      # Twiddle factors;
        Wj = 1.
        for j in range(N1):                    # Compute every subsequence of size N2;
            Xj = recursive_fft(x[j::N1])       # Recursively compute the Fourier Transform;
            Wkj = 1.
            for k in range(N):
                X[k] = X[k] + Xj[k%N2] * Wkj   # Recombine results;
                Wkj = Wkj * Wj                 # Update twiddle factors;
            Wj = Wj * W
        return X


####################################################################################################
# Recursive FFT:
def vec_recursive_fft(x):
    """
    Fast Fourier Transform using a recursive decimation in time algorithm. This has smaller
    complexity than the direct FT, though the exact value is difficult to compute. This
    implementation uses NumPy arrays for conciseness. In this implementation, loops are avoided by
    vectorizing the computation of the twiddle factors.

    :Parameters:
      x
        The vector of which the FFT will be computed. It must be a composite number, or else the
        computation will be defered to the direct FT, and there will be no efficiency gain.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of the vector;
    N1 = __factor(N)                           # Find the smallest factor of the vector length;
    if N1 == N:                                # If the length is prime itself,
        return direct_ft(x)                    #    the transform is given by the direct form;
    else:
        N2 = N // N1                           # Decompose in two factors, N1 being prime;
        X = zeros((N, ), dtype=complex)        # Accumulate the results;
        k = arange(N)
        for j in range(N1):                    # Compute every subsequence of size N2;
            Xj = vec_recursive_fft(x[j::N1])   # Recursively compute the Fourier Transform;
            Wkj = exp(-2j*pi*k*j/N)
            X = X + Xj[k%N2]*Wkj         # Recombine results;
        return X


####################################################################################################
# Main program:
if __name__ == "__main__":

    SIZES = [ 2*3, 2*2*3, 2*3*3, 2*3*5, 2*2*3*3, 2*2*5*5, 2*3*5*7, 2*2*3*3*5*5 ]

    # Start printing the table with time comparisons:
    print("+---------"*6 + "+")
    print("|    N    |   N^2   | Direct  | Recurs. | VecRec. | Intern. |")
    print("+---------"*6 + "+")

    # Try it with vectors with the given sizes:
    for n in SIZES:

        # Compute the average execution time:
        dtime  = time_it(direct_ft, n, REPEAT)
        rtime  = time_it(recursive_fft, n, REPEAT)
        vtime  = time_it(vec_recursive_fft, n, REPEAT)
        intime = time_it(fft.fft, n, REPEAT)

        # Print the results:
        tup = (n, n**2, dtime, rtime, intime)
        print(f'| {n:7} | {n**2:7} | {dtime:7.4f} | {rtime:7.4f} | {vtime:7.4f} | {intime:7.4f} |')

    print("+---------"*6 + "+")

File: ocaml/anyfft_array.ml
(**************************************************************************************************
 Fast Fourier Transform -- OCaml Version
 This version implements Cooley-Tukey algorithm for composite numbers (not powers of 2 only) using
 OCaml arrays.

 José Alexandre Nalon
 **************************************************************************************************
 There are different ways to compile and run an OCaml program. To run directly from the command
 line, just type:

 $ ocaml anyfft_array.ml

 And the program will run. To compile the program, use the command:

 $ ocamlc -o anyfft_array anyfft_array.ml

 This will generate an executable file named `anyfft_array` in the same directory, that can be run
 from the command line by typing:

 $ ./anyfft_array
 **************************************************************************************************)

(**************************************************************************************************
 Open needed modules:
 **************************************************************************************************)
open Array                             (* Array of numbers; *)
open Complex                           (* Complex numbers; *)
open Printf                            (* Output results; *)
open Sys                               (* Time measurement; *)


(**************************************************************************************************
 Definitions:
 **************************************************************************************************)
let repeats = 50
let pi = 3.1415926535897931


(**************************************************************************************************
 Auxiliary Functions:
 **************************************************************************************************)
(* Convert integer to complex. *)
let complex_of_int n = { re=(float_of_int n); im=0. }


(* Integer power of a complex number. *)
let cpow w n = Complex.pow w (complex_of_int n)


(* Creates a sequence of integer numbers. *)
let range a b = List.init (b - a + 1) (fun n -> n + a)


(**************************************************************************************************
 Auxiliary function: complex_show
   Pretty printing of an array of complex numbers, used to inspect results.

 Parameters:
   x
     An array of complex numbers;
 **************************************************************************************************)
let complex_show x =
    for i = 0 to ((length x) - 1) do
        printf "( %7.4f, %7.4f )\n" x.(i).re x.(i).im
    done


(**************************************************************************************************
 Measure execution time through repeated calls to a (Fast) Fourier Transform function.

 Parameters:
   f
     Function to be called;
   size
     Power of two of the size of the vector on which the transform will be applied;
   repeats
     Number of times the function will be called.

 Returns:
   The average execution time for that function with a vector of the given size.
 **************************************************************************************************)
let time_it f size repeats =
    let x = Array.init size (fun n -> { re=(float n); im=0. }) in
    let t0 = Sys.time () in
    for i = 1 to size do
        f x;
    done;
    (Sys.time () -. t0) /. (float repeats)


(**************************************************************************************************
 Discrete Fourier Transform directly from the definition, an algorithm that has O(N^2) complexity.

 Parameters:
   x
     The vector of which the DFT will be computed. Given the nature of the implementation, there is
     no restriction on the size of the list, although it will almost always be called with a power
     of two size to give a fair comparison.

 Returns:
     A complex-number vector of the same size, with the coefficients of the DFT.
 **************************************************************************************************)
let direct_ft x =
    let nx = length x in
    let tX = Array.init nx (fun n -> Complex.zero) in
    let w = exp { re=0.; im= -2. *. pi /. (float nx) } in
    let wk = ref { re=1.; im=0. } in
    for k = 0 to (nx - 1) do
        let wkn = ref { re=1.; im=0. } in
        for n = 0 to (nx - 1) do
            tX.(k) <- add tX.(k) (mul !wkn x.(n));
            wkn := mul !wkn !wk
        done;
        wk := mul !wk w
    done;
    tX


(**************************************************************************************************
 Smallest prime factor of a given number. If the argument is prime itself, then it is the return
 value.

 Parameters:
   n
     Number to be inspected;

 Returns:
   The smallest prime factor, or the number itself if it is already a prime.
 **************************************************************************************************)
let factor n =
    let rec loop n k =
        if n mod k == 0 then
            k
        else if k == n / 2 then
            n
        else
            loop n (k + 1)
    in loop n 2


(**************************************************************************************************
 Fast Fourier Transform using a recursive decimation in time algorithm. This has O(N log_2(N))
 complexity.

 Parameters:
   x
     The vector of which the FFT will be computed. Its length must be a composite number, or else
     the computation will be defered to the direct FT, and there will be no efficiency gain.

 Returns:
   A complex-number vector of the same size, with the coefficients of the DFT.
 **************************************************************************************************)
let rec recursive_fft x =
    let nx = Array.length x in
    let n1 = factor nx in
    if n1 == nx then
        direct_ft x
    else
        let n2 = nx / n1 in
        let xj = Array.make n2 Complex.zero in
        let tX = Array.make nx Complex.zero in
        let w = exp { re=0.; im= -2. *. pi /. (float nx) } in
        let wj = ref { re=1.; im=0. } in
        for j = 0 to (n1 - 1) do
            for k = 0 to (n2 - 1) do
                xj.(k) <- x.(k * n1 + j)
            done;
            let tXj = recursive_fft xj in
            let wkj = ref { re=1.; im=0. } in
            for k = 0 to (nx - 1) do
                tX.(k) <- add tX.(k) (mul tXj.(k mod n2) !wkj);
                wkj := mul !wkj !wj
            done;
            wj := mul !wj w
        done;
        tX


(**************************************************************************************************
 Main function:
 **************************************************************************************************)
let main () =

    (* Start by printing the table with time comparisons: *)
    printf "+---------+---------+---------+---------+\n";
    printf "|    N    |   N^2   | Direta  | Recurs. |\n";
    printf "+---------+---------+---------+---------+\n";

    (* Try it with vectors with the given sizes: *)
    let sizes = [ 2*3; 2*2*3; 2*3*3; 2*3*5; 2*2*3*3; 2*2*5*5; 2*3*5*7; 2*2*3*3*5*5 ] in
    let main_loop n =

        (* Compute the average execution time: *)
        let dtime = time_it direct_ft n repeats in
        let rtime = time_it recursive_fft n repeats in

        (* Print the results: *)
        printf "| %7d | %7d | %7.4f | %7.4f |\n" n (n*n) dtime rtime;
        flush stdout

    in List.iter main_loop sizes;

    printf "+---------+---------+---------+---------+\n"

let () = main ()
